libtenzir/builtins/operators/chart2.cpp

```diff
@@ -0,0 +1,794 @@
+      TENZIR_ASSERT(ptr);
+      plugins.emplace(ptr, std::move(wrapped_call));
+    }
+    return plugins;
+  }
+
+  auto make_bucket(const plugins_map& plugins, session ctx) const -> bucket {
+    auto b = bucket{};
+    for (const auto& [plugin, arg] : plugins) {
+      auto inv = aggregation_plugin::invocation{arg};
+      auto instance = plugin->make_aggregation(std::move(inv), ctx);
+      TENZIR_ASSERT(instance);
+      b.push_back(std::move(instance).unwrap());
+    }
+    return b;
+  }
+};
+
+class chart_operator2 final : public crtp_operator<chart_operator2> {
+public:
+  chart_operator2() = default;
+
+  explicit chart_operator2(chart_args args) : args_{std::move(args)} {
+  }
+
+  auto name() const -> std::string override {
+    return "tql2.chart";
+  }
+
+  auto optimize(const expression& filter, event_order) const
+    -> optimize_result override {
+    // FIXME: Handle this properly
+    auto expr = filter;
+    if (args_.xmin) {
+      expr = conjunction{args_.xmin->legacy_expr, std::move(expr)};
+    }
+    if (args_.xmax) {
+      expr = conjunction{args_.xmax->legacy_expr, std::move(expr)};
+    }
+    auto combined = normalize_and_validate(expr);
+    TENZIR_ASSERT(combined);
+    // NOTE: This should technically be `ordered` but since most of our useful
+    // aggregations currently are commutative, we can get away with this.
+    return {std::move(combined).value(), event_order::unordered, copy()};
+  }
+
+  auto
+  operator()(generator<table_slice> input, operator_control_plane& ctrl) const
+    -> generator<table_slice> {
+    // Using a `deque` to guarantee reference validity after growing
+    auto ynums = std::deque<std::string>{"y"};
+    auto gnames = std::unordered_set<std::string>{};
+    auto xpath = args_.x.path()[0].name;
+    for (auto i = size_t{1}; i < args_.x.path().size(); ++i) {
+      xpath += ".";
+      xpath += args_.x.path()[i].name;
+    }
+    auto& dh = ctrl.diagnostics();
+    auto sp = session_provider::make(dh);
+    auto s = sp.as_session();
+    auto xty = std::optional<type>{};
+    auto groups = group_map{};
+    const auto plugins = args_.find_plugins(s);
+    const auto* null_group = get_groups(groups, caf::none, s);
+    if (args_.xmin or args_.xmax) {
+      input = filter(std::move(input), dh);
+    }
+    for (auto&& slice : input) {
+      if (slice.rows() == 0) {
+        co_yield {};
+        continue;
+      }
+      auto consumed = size_t{};
+      auto xs = eval(args_.x, slice, dh);
+      auto gs = get_group_strings(slice, dh);
+      TENZIR_ASSERT(gs.type.kind().is<string_type>());
+      if (not xty) {
+        if (not validate_x(xs.type, dh)) {
+          consumed += xs.length();
+          continue;
+        }
+        xty = xs.type;
+      }
+      if (xs.type != xty.value()) {
+        diagnostic::warning("cannot plot different types `{}` and `{}` on "
+                            "the X-axis",
```

Comment:
```suggestion
                            "the x-axis",
```
same for the other occurrences